feat: Add rate limiting to identity search endpoint

srdasdev opened the pull request:
- **Backend**: Add `ScopedRateThrottle` to `IdentityViewSet` (30 req/min)
- **Frontend**: Increase debounce from 500ms to 750ms
- **Config**: Add `IDENTITY_SEARCH_THROTTLE_RATE` env variable
- **Tests**: Add test for identity search throttling

This fixes aggressive API requests when searching for identities.

---

Thanks for submitting a PR! Please check the boxes below:

- [ ] I have added information to `docs/` if required so people know about the feature
- [x] I have filled in the **Changes** section below
- [x] I have filled in the **How did you test this code** section below
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

---

## Changes

### Backend Changes
- **`api/environments/identities/views.py`**  
  Added `ScopedRateThrottle` to `IdentityViewSet` with scope `identity_search`

- **`api/app/settings/common.py`**  
  Added configurable `IDENTITY_SEARCH_THROTTLE_RATE` environment variable (default: `30/min`)

### Frontend Changes
- **`frontend/common/useDebouncedSearch.ts`**  
  Increased debounce time from 500ms to 750ms to reduce request frequency during typing

### Configuration

The rate limit can be customized via environment variable:

```bash
IDENTITY_SEARCH_THROTTLE_RATE="30/min"  # default
````

---

## How did you test this code?

### Unit Tests

* Added `test_identity_search_is_throttled` in
  **`api/tests/unit/environments/identities/test_unit_identities_views.py`**

### Manual Testing

1. Built a local Docker image with the changes
2. Made 35 rapid requests to the identity search endpoint:

```bash
for i in {1..35}; do
  curl -s -o /dev/null -w "Request $i: %{http_code}\n" \
    -H "Authorization: Token $TOKEN" \
    "http://localhost:8009/api/v1/environments/$ENV_KEY/identities/?q=test"
done
```

3. Verified that requests after the 30/min threshold return **429 Too Many Requests**:

```
Request 33: 200
Request 34: 429  ← Rate limit triggered
Request 35: 429
```

```
```

smy-637q committed:
feat: Add rate limiting to identity search endpoint

- Backend: Add ScopedRateThrottle to IdentityViewSet (30 req/min)
- Frontend: Increase debounce from 500ms to 750ms
- Config: Add IDENTITY_SEARCH_THROTTLE_RATE env variable
- Tests: Add test for identity search throttling

Fixes aggressive API requests when searching for identities.

smy-637q committed:
fix: Apply identity search throttle only to list action

- Use get_throttles() method to conditionally apply throttle only to 'list' action
- Update test to use mocker.patch and reset_cache fixture for proper cleanup
- Follow existing patterns from FFAdminUserViewSet and OrganisationViewSet

srdasdev committed:
Merge branch 'Flagsmith:main' into feature/identity-search-rate-limiting

smy-637q committed:
fix: Add identity_search throttle rate to test settings

The test.py settings file overrides DEFAULT_THROTTLE_RATES from common.py,
so identity_search scope was missing in the test environment.

srdasdev committed:
Merge branch 'Flagsmith:main' into feature/identity-search-rate-limiting

matthewelwell commented:
> > @1-23-smy could you check the unit test failures here please?
> 
> @matthewelwell Could you review once ?

Thanks - I'll take a look, but you do have some conflicts that also need resolving please.

srdasdev committed:
Merge branch 'feature/identity-search-rate-limiting' of https://github.com/1-23-smy/flagsmith

1-23-smy commented:
> > > @1-23-smy could you check the unit test failures here please?
> > 
> > 
> > @matthewelwell Could you review once ?
> 
> Thanks - I'll take a look, but you do have some conflicts that also need resolving please.

@matthewelwell I have fixed the conflicts. Could you please review once ?

Zaimwa9 reviewed (changes requested):
Thanks for the contribution 👍 . 2 comments, the most important one would be to stick with the previous debounce time if you don't mind

smy-637q committed:
fix: Address reviewer feedback

- Revert frontend debounce from 750ms back to 500ms (agreed UX time)
- Return super().get_throttles() for non-list actions so global default throttle classes still apply

1-23-smy commented on api/environments/identities/views.py:
```diff
@@ -41,6 +42,15 @@
+        """
+        if getattr(self, "action", None) == "list":
+            return [ScopedRateThrottle()]
+        return []
```
@Zaimwa9 

- I have reverted debounce from 750ms back to 500ms (agreed UX time).
- Updated get_throttles() to return super().get_throttles() for non-list actions.

1-23-smy replied:
Thanks for the review! the conflicts should now be resolved.

Let me know if there's anything else you'd like me to adjust!

smy-637q committed:
Merge branch 'main' of https://github.com/1-23-smy/flagsmith

Zaimwa9 committed:
feat: re-added-master-api-key-throttle

Zaimwa9 committed:
Merge branch 'main' of github.com:Flagsmith/flagsmith into feature/identity-search-rate-limiting

Zaimwa9 committed:
feat: re-added-deleted-values

Zaimwa9 reviewed (dismissed):
Looking good to me, I allowed myself to fix some loss - probably while fixing conflicts. Thanks for the contribution!

srdasdev reviewed (commented):
Thank You.

srdasdev committed:
Merge branch 'main' into feature/identity-search-rate-limiting
